lorawan: handling of datarate changes

JordanYates opened the pull request:
When enabling Adaptive-Data-Rate (ADR), the datarate of the network connection changes is response to packets transmitted and received. As datarates change, so too does the payload capacity of a single packet. What once was a valid payload may no longer fit in a single packet, or several packets may be able to be consolidated due to an increase in payload size, improving network efficiency. It is therefore useful to know both how much data we can possibly transmit, and to be notified when that value changes. 

This functionality also allows devices to handle the rare but catastrophic case of a network server reboot. As time passes without receiving acknowledgements, the ADR algorithm will automatically reduce the datarate used. However it will only reduce the datarate to the minimum possible datarate for a given region. Once there, no further action is taken, resulting in a node permanently stuck with the wrong network parameters. The correct action is to assume the network is lost if packets are still not being acknowledged at the lowest datarate, and initiate a rejoin.

This PR addresses the above problems. Firstly functionality to query possible payload sizes is added. Secondly a callback is provided to be notified when the datarate is changed. Finally, the minimum datarate can be queried.

Commit:
lorawan: queriable payload sizes

Add the ability for applications to query the maximum size of packets
that can be sent. This must be dynamically queried as the sizes change
with datarate, region, and as MAC commands are added by the stack.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lorawan: store current datarate for ADR

Keep track of the current network datarate when ADR is enabled.

The datarate can change in two circumstances. Firstly, the datarate may
change as a result of a command from the LoRaWAN server as the link
budget varies. Secondly, the datarate may be changed due to not
receiving an expected ADRACKReq response. This necessitates querying the
datarate on both packet reception and transmission to provide timely
notifications to the application.

Due to querying the default region datarate at startup and validating
manual datarate parameters, when ADR is not enabled the datarate will
never be different from the value provided to `lorawan_set_datarate`.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lorawan: queriable minimum datarate

Add a function to query the minimum possible datarate on the network.
This value may change over the lifetime of the connection as a result
of commands from the network server.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lorawan: add callback for datarate changes

Add a callback to notify applications when the datarate has changed.
This allows applications to know when payload sizes have changed without
needing to call `lorawan_get_payload_sizes` before each transmission.

This also enables:
 * Monitoring of network conditions on the device
 * Determining if a network connection has been lost

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
samples: lorawan: class_a: add datarate callback

Add an example callback implementation that prints out the new maximum
payload given the updated datarate.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

tagunil commented:
As far as I can see, the interface for querying various MAC parameters is LoRaMacMibGetRequestConfirm. Wouldn't it be better to extend that API and to propose such changes to LoRaMac-node upstream?

JordanYates commented:
> Wouldn't it be better to extend that API and to propose such changes to the LoRaMac-node upstream?

Of course, I was under the impression that the upstream repo was code distributed by Semtech, not accepting PR's.
This is of course incorrect, and I have submitted a patch (https://github.com/Lora-net/LoRaMac-node/pull/1029).
This patch cannot be applied directly to the Zephyr fork, so I will submit a variant.

Commit:
manifest: update loramac-node module

Update the loramac-node module to a commit containing the ability to
query the current minimum datarate.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JayKickliter commented:
> What once was a valid payload may no longer fit in a single packet, or several packets may be able to be consolidated due to an increase in payload size

This is specifically why Semtech's recommended ADR algorithm, for all its faults, will only ever increase datarate and **never decrease it**. I recently implemented ADR for another LoRaWAN network sever and did the same, but I can't speak for every LNS.

JordanYates commented:
> This is specifically why Semtech's recommended ADR algorithm, for all its faults, will only ever increase datarate and **never decrease it**. I recently implemented ADR for another LoRaWAN network server and did the same, but I can't speak for every LNS.

From the point of view of the network server this is correct, however not from the view of the complete system.
The network server will never downgrade a datarate, but the node itself will downgrade it upon failing to hear packets.

Every `{REGION}_ADR_ACK_LIMIT` packets without receiving a downlink, the  node will set the `ADRACKReq` bit in the TX frame.
The network server then has `{REGION}_ADR_ACK_DELAY` further packets to transmit a downlink packet to the node.
If the node does not hear a downlink packet in that time, it will reduce the datarate.

This is the default behaviour of `loramac-node`, and can be easily observed with this PR by turning off your gateway and watching the datarate drop.

brocaar commented:
> Maybe @brocaar can chime in on how Chirp Stack does it. Maybe it's not an actual concern?

ChirpStack's ADR algorithm only increases the DR, it never decreases the DR. Note that from the server-side, the max payload size will always be a challenge. The AS may enqueue multiple payloads with the max payload size. But there is no guarantee that when a downlink opportunity occurs, the device is still using the same DR. Potentially, the device lowered its DR. In this case when the downlink payload exceeds the max payload limit, ChirpStack will notify the AS that it dropped the item from the downlink queue and it will try to enqueue the next item in the queue.

Mani-Sadhasivam commented on subsys/lorawan/lorawan.c:
```diff
@@ -429,14 +455,14 @@ int lorawan_send(uint8_t port, uint8_t *data, uint8_t len, uint8_t flags)
 			mcpsReq.Req.Confirmed.fBufferSize = len;
 			mcpsReq.Req.Confirmed.NbTrials = lorawan_conf_msg_tries;
-			mcpsReq.Req.Confirmed.Datarate = lorawan_datarate;
+			mcpsReq.Req.Confirmed.Datarate = default_datarate;
```
Why can't this be `current_datarate`?

JordanYates replied:
It doesn't matter whether it is current or default datarate.
If ADR is enabled, this value is ignored.
If ADR is disabled, `current_datarate` is set to `default_datarate` due to the logic at the end of `lorawan_join`.

Mani-Sadhasivam replied:
Right, but using `default_datarate` doesn't sound right to me. But anyway, I was wondering if we can just use the single variable for caching the datarate.

JordanYates replied:
I can change the name over. As for merging `default_datarate` and `current_datarate`, that depends on what the intended behavior of the following code is:
```
int main(void) {
    lorawan_enable_adr(true);
    lorawan_set_datarate(DR_2);
    lorawan_join(&join_cfg);
    for(int i = 0; i < 50; i++) {
        lorawan_send(data, data_len);
    }
    /* Rejoin the network */
    lorawan_join(&join_cfg);
}
```
If the second join request should use the same datarate that we set originally, we need to preserve that value.
Keep in mind that the second join could be to a completely different network, with different link budgets etc.
I don't actually want to join multiple networks, but other people might.

Mani-Sadhasivam replied:
> If the second join request should use the same datarate that we set originally, we need to preserve that value.
Keep in mind that the second join could be to a completely different network, with different link budgets etc.
I don't actually want to join multiple networks, but other people might.

Hmm, makes sense. Then please use `current_datarate` for `lorawan_send` and add a note above the declaration of `default_datarate` on why we have 2 variables. Otherwise LGTM!

Mani-Sadhasivam reviewed (approved):
LGTM!

Mani-Sadhasivam commented:
@carlescufi @nashif Please merge!
